### PR TITLE
feat(associated token pda search): add tokenProgramId param

### PR DESCRIPTION
**Pull request by shotgunofdeath**
# Description:
`findAssociatedTokenPda()` is updated. Now it accepts another seed - `tokenProgramId`. It’s a specific public key for either Token or Token2022 Program as we discussed with @blockiosaurus. Tests are updated accordingly.

NB! JS tests will fail because `findAssociatedTokenPda()` is used in generated by Kinobi instructions. Specifically:
src/generated/instructions/createAssociatedToken.ts
src/generated/instructions/createTokenIfMissing.ts

And generated files shouldn’t be changed manually.

BREAKING CHANGE: the `tokenProgramId` param for the `findAssociatedTokenPda()` function breaks js tests which use instructions generated by Kinobi. It will also break any code relying on this function.

### CODE REVIEW

**Comment by shotgunofdeath** [comment]
Not sure if it’s correct or necessary. But I have some suggestions to simplify the usage of `findAssociatedTokenPda()` for end users. We can make it accept not a specific `PublicKey` type but either a flag with a default value or string with a token program name for umi and find the necessary public key in the function itself. E.g.

```
isToken2022: boolean = false
const tokenProgramId = context.programs.getPublicKey(
    isToken2022 ? 'splToken2022' : 'splToken'
  );
```

or

```
tokenProgramIdVersion: 'splToken' | 'splToken2022' = 'splToken'
const tokenProgramId = context.programs.getPublicKey(tokenProgramVersion)
```

Also I’m not sure whether specific `PublicKeys` types and constants for token programs should be defined in mpl-toolbox. It seems it’s better to have them on some another level, e.g. in umi itself.

**Comment by lorisleiva** [comment]
Hi there, have you seen this?

https://solana.stackexchange.com/a/9121/1668

**Comment by shotgunofdeath** [comment]
Hello there! No, not before you've shown this.

File: clients/js/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@metaplex-foundation/mpl-toolbox",
-  "version": "0.9.2",
+  "version": "0.9.3",
```
Comment (blockiosaurus) [inline clients/js/package.json:3]: I think we want this to be 0.10.0 so downstream packages don't pull these changes.

File: clients/js/src/hooked/AssociatedToken.ts [hunk 2]
```diff
@@ -8,14 +9,15 @@ export function findAssociatedTokenPda(
     mint: PublicKey;
     /** The owner of the token account */
     owner: PublicKey;
+    /** The Token or Token2022 Program id */
```
Comment (blockiosaurus) [inline clients/js/src/hooked/AssociatedToken.ts:12]: Yes, I like your idea of using strings for token program name. Let's do that!

Comment (shotgunofdeath) [reply clients/js/src/hooked/AssociatedToken.ts]: https://github.com/metaplex-foundation/mpl-toolbox/pull/12/commits/e8306418ea63a8dd4003419e8691a83719f95eb9

File: clients/js/package.json (hunk 1 above)
Comment (shotgunofdeath) [reply clients/js/package.json]: https://github.com/metaplex-foundation/mpl-toolbox/commit/e8306418ea63a8dd4003419e8691a83719f95eb9

**Comment by lorisleiva** [comment]
My point before was that this manually written function follows the convention of what is generated by Kinobi and therefore allows devs to swap program via Umi's program repository.

This change would break that convention and that function would now behave differently then the rest of the Umi's PDA helpers which I wouldn't recommend.

If you don't like the idea of using the program repository, I would at least suggest accepting the token program as an input and, when not provided, resolving it from Umi's program repository as a fallback so this doesn't introduce a breaking chance.

**Comment by blockiosaurus** [comment]
> My point before was that this manually written function follows the convention of what is generated by Kinobi and therefore allows devs to swap program via Umi's program repository.
> 
> This change would break that convention and that function would now behave differently then the rest of the Umi's PDA helpers which I wouldn't recommend.
> 
> If you don't like the idea of using the program repository, I would at least suggest accepting the token program as an input and, when not provided, resolving it from Umi's program repository as a fallback so this doesn't introduce a breaking chance.

Yes but doing it via the program repository wouldn't work for programs using both SPL token programs at the same time. Token22 set the pretend of spl-token no longer being special, which this change reflects.

**Comment by lorisleiva** [comment]
You could bind and unbind afterwards but I understand that having an explicit input might be better for Devex. However, if the token program is not provided explicitly, we should fallback to resolving the program via the repository so that this change isn't a breaking one and so that this function is still consistent with the rest of the generated function.

**Comment by shotgunofdeath** [comment]
@blockiosaurus I've made some changes according to @lorisleiva recommendations. Let me, please, know if it's OK with you both, guys.

https://github.com/metaplex-foundation/mpl-toolbox/pull/12/commits/7f33d3b529172687a2886887c405138167b5f7d6

File: clients/js/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@metaplex-foundation/mpl-toolbox",
-  "version": "0.9.2",
+  "version": "0.10.0",
```
Comment (lorisleiva) [inline clients/js/package.json:3]: I think a patch bump is okay now since we're no longer adding a breaking change.

**Review by lorisleiva — APPROVED** [review]
Much better, thanks!

**Review by blockiosaurus — APPROVED** [review]
This looks great, thank you!

**Comment by shotgunofdeath** [comment]
> This looks great, thank you!

Thank you! 

@blockiosaurus @lorisleiva Appreciate all the help!
I've squashed the fixup commits. I guess you could merge and publish now.

**Pull request merged**